### PR TITLE
GBT inference optimization

### PR DESCRIPTION
**Pull request by razdoburdin**
Improving the performance of GBT inference by making a block_size flexible. The speedup is up to 2x.
For performance benchmarking I used [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) for xgboost/lightgbm/catboost modelbuilders.

### CODE REVIEW

File: cpp/daal/src/algorithms/dtrees/gbt/regression/gbt_regression_predict_dense_default_batch_impl.i [hunk 1]
```diff
@@ -61,6 +60,7 @@ class PredictRegressionTask
 {
 public:
     typedef gbt::internal::GbtDecisionTree TreeType;
+    typedef gbt::prediction::internal::TileDimensions<algorithmFPType> DimType;
```
Comment (Vika-F) [inline cpp/daal/src/algorithms/dtrees/gbt/regression/gbt_regression_predict_dense_default_batch_impl.i:63]: Can you please look at the approach to block size calculation used in K-means: https://github.com/oneapi-src/oneDAL/blob/master/cpp/daal/src/algorithms/kmeans/kmeans_lloyd_batch_impl.i#L140

DAAL_SAFE_CPU_CALL macro sets up default block size for third-party CPUs.

File: cpp/daal/src/algorithms/dtrees/gbt/gbt_predict_dense_default_impl.i [hunk 2]
```diff
@@ -155,24 +156,36 @@ struct TileDimensions
...
-            if (daal::threader_get_threads_number() > nRowsTotal / nRowsInBlock)
-            {
-                nRowsInBlock = VECTOR_BLOCK_SIZE;
-            }
-        }
-        nDataBlocks = nRowsTotal / nRowsInBlock;
+        // Use smaller vectorBlockSize if trees fit to L2
+        // Each node contain 3 values
+        size_t nodesSize                = (sizeof(ModelFPType) + sizeof(FeatureIndexType) + sizeof(int)) * nNodes;
+        constexpr float add_hoc_L2_size = 0.85 * 2 * 1024 * 1024;
```
Comment (Vika-F) [inline cpp/daal/src/algorithms/dtrees/gbt/gbt_predict_dense_default_impl.i:177]: Please use daal::services::internal::getL2CacheSize() to define the size of L2 cache.
There are also other functions you might find useful in "service_envirinment.h"

**Review by Vika-F — COMMENTED** [review]
The changes are good, but it would be better to implement more general solution based on exact cache sizes and taking into account third-party CPUs as well.

File: cpp/daal/src/algorithms/dtrees/gbt/regression/gbt_regression_predict_dense_default_batch_impl.i (hunk 1 above)
Comment (razdoburdin) [reply cpp/daal/src/algorithms/dtrees/gbt/regression/gbt_regression_predict_dense_default_batch_impl.i:63]: done

File: cpp/daal/src/algorithms/dtrees/gbt/gbt_predict_dense_default_impl.i (hunk 2 above)
Comment (razdoburdin) [reply cpp/daal/src/algorithms/dtrees/gbt/gbt_predict_dense_default_impl.i]: done

File: cpp/daal/src/algorithms/dtrees/gbt/gbt_predict_dense_default_impl.i [hunk 3]
```diff
@@ -155,24 +156,36 @@ struct TileDimensions
...
-
-            if (daal::threader_get_threads_number() > nRowsTotal / nRowsInBlock)
-            {
-                nRowsInBlock = VECTOR_BLOCK_SIZE;
-            }
-        }
-        nDataBlocks = nRowsTotal / nRowsInBlock;
+        // Use smaller vectorBlockSize if trees fit to L2
+        // Each node contain 3 values
+        size_t nodesSize        = (sizeof(ModelFPType) + sizeof(FeatureIndexType) + sizeof(int)) * nNodes;
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/dtrees/gbt/gbt_predict_dense_default_impl.i:176]: Can you estimate this size just by calling `sizeof(NodeType)`?

File: cpp/daal/src/algorithms/dtrees/gbt/gbt_predict_dense_default_impl.i [hunk 4]
```diff
@@ -155,24 +156,36 @@ struct TileDimensions
...
+
+        // Decrease vectorBlockSize if number of rows is too small
+        const size_t twoBlocksPerThreadFactor = nRowsTotal / (2 * daal::threader_get_threads_number() * vectorBlockSizeStep);
+        if (vectorBlockSizeFactor > twoBlocksPerThreadFactor) vectorBlockSizeFactor = twoBlocksPerThreadFactor;
+        if (vectorBlockSizeFactor < minVectorBlockSizeFactor) vectorBlockSizeFactor = minVectorBlockSizeFactor;
+
+        DAAL_SAFE_CPU_CALL({}, vectorBlockSizeFactor = optimalBlockSizeFactor)
+        size_t vectorBlockSize = vectorBlockSizeStep * vectorBlockSizeFactor;
+        nRowsInBlock           = nRowsTotal > vectorBlockSize ? vectorBlockSize : nRowsTotal;
+        nDataBlocks            = nRowsTotal / nRowsInBlock + (nRowsTotal % nRowsInBlock != 0);
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/dtrees/gbt/gbt_predict_dense_default_impl.i:188]:
```suggestion
        nDataBlocks            = nRowsTotal / nRowsInBlock + bool(nRowsTotal % nRowsInBlock);
```

File: cpp/daal/src/algorithms/dtrees/gbt/gbt_predict_dense_default_impl.i (hunk 3 above)
Comment (razdoburdin) [reply cpp/daal/src/algorithms/dtrees/gbt/gbt_predict_dense_default_impl.i:176]: Unfortunately, but the [corresponding structure](https://github.com/razdoburdin/oneDAL/blob/ec6b6c593dc5241bfd82e5aa72d5721b9d0f5a39/cpp/daal/src/algorithms/dtrees/dtrees_model_impl.h#LL50C1-L50C1) contain extra fields.

**Comment by razdoburdin** [comment]
Private CI finished with warnings. Public CI errors are not related to the PR.

**Review by Alexsandruss — CHANGES_REQUESTED** [review]
Rebase is required to run PreCommit because of significant new commits in oneDAL/sklearnex repos

**Comment by razdoburdin** [comment]
> Rebase is required to run PreCommit because of significant new commits in oneDAL/sklearnex repos

done

**Review by Alexsandruss — APPROVED** [review]

**Pull request merged**